### PR TITLE
Reference API DF support

### PR DESCRIPTION
**Pull request by qx-teo**
### Description
Takes missing dates from API dataframe, and supplements it with test data for days before the checking date. Useful for non-daily updates (or large batch updates in event of outages)

### Changelog
Itemize code/test/documentation changes and files added/removed.
- dynamic.py

### Fixes 
- Fixes #(issue)

### CODE REVIEW

**Comment by qx-teo** [comment]
@nmdefries

File: _delphi_utils_python/delphi_utils/validator/dynamic.py [hunk 1]
```diff
@@ -192,6 +192,8 @@ def validate(self, all_frames, report):
                                           "be performed"))
                     continue
 
+                self.check_reference_api_df(reference_api_df, geo_sig_df, reference_end_date)
```
Comment (nmdefries) [inline _delphi_utils_python/delphi_utils/validator/dynamic.py:195]: This isn't actually saving the returned padded ref API data anywhere, so subsequent checks will still use the un-padded reference data.

File: _delphi_utils_python/delphi_utils/validator/dynamic.py [hunk 2]
```diff
@@ -254,6 +256,32 @@ def check_max_allowed_max_date(self, max_date, geo_type, signal_type, report):
 
         report.increment_total_checks()
 
+    def check_reference_api_df(self, reference_api_df, geo_sig_df, reference_end_date):
```
Comment (nmdefries) [inline _delphi_utils_python/delphi_utils/validator/dynamic.py:259]: Looks like this isn't actually checking anything, so a different name would be more suitable, e.g. "pad_to_date" or something.

**Review by nmdefries — CHANGES_REQUESTED** [review]

File: _delphi_utils_python/delphi_utils/validator/dynamic.py [hunk 3]
```diff
@@ -254,6 +256,32 @@ def check_max_allowed_max_date(self, max_date, geo_type, signal_type, report):
...
+        if reference_api_df_max_date < reference_end_date:
+            # Querying geo_sig_df, only taking relevant rows
+            geo_sig_df_supplement = geo_sig_df.query(
+                'time_value <= @reference_end_date & time_value > \
+                @reference_api_df_max_date')[[
+                "geo_id", "val", "se", "sample_size", "time_value"]]
+            # Matching time_value format
+            geo_sig_df_supplement["time_value"] = \
+                pd.to_datetime(geo_sig_df_supplement["time_value"],
+                    format = "%Y-%m-%d %H:%M:%S")
```
Comment (nmdefries) [inline _delphi_utils_python/delphi_utils/validator/dynamic.py:280]: This logic is fairly complicated (easy to have an off-by-one error on the dates, for example) so it'd good to add a few tests.

File: _delphi_utils_python/delphi_utils/validator/dynamic.py (hunk 1 above)
Comment (qx-teo) [reply _delphi_utils_python/delphi_utils/validator/dynamic.py:195]: Good catch! I'll add it in

**Review by nmdefries — APPROVED** [review]

**Pull request merged**